### PR TITLE
Issue/194

### PR DESCRIPTION
**Pull request by NoCtrlZ**
Hi @ed255 
I drafted `eth_getCode` method issue https://github.com/appliedzkp/zkevm-circuits/issues/194.
I also setup the test env with `ganache` which supports same RPC with geth.
By indicating same mnemonic, we can produce same contract addresses and block hashes.
You can check my code with `$ docker-compose up` and `cd bus-mapping && cargo test -- --ignored`.

Thank you!

### CODE REVIEW

File: bus-mapping/src/rpc.rs [hunk 1]
```diff
@@ -119,6 +119,17 @@ impl<P: JsonRpcClient> GethClient<P> {
             .map_err(|e| Error::JSONRpcError(e.into()))?;
         Ok(resp.0.into_iter().map(|step| step.result).collect())
     }
+
+    /// Calls `eth_getCode` via JSON-RPC returning a contract code
+    pub async fn get_code_by_address(
+        &self,
+        contract_address: String,
```
Comment (spartucus) [inline bus-mapping/src/rpc.rs:126]: nitpick: For contract address, use `Address` instead of `String` would be better, what do you think?

Comment (NoCtrlZ) [reply bus-mapping/src/rpc.rs:126]: Thank you for the review!
I think that's more beautiful.
I am going to change.

File: bus-mapping/src/rpc.rs [hunk 2]
```diff
@@ -119,6 +119,20 @@ impl<P: JsonRpcClient> GethClient<P> {
             .map_err(|e| Error::JSONRpcError(e.into()))?;
         Ok(resp.0.into_iter().map(|step| step.result).collect())
     }
+
+    /// Calls `eth_getCode` via JSON-RPC returning a contract code
+    pub async fn get_code_by_address(
+        &self,
+        contract_address: Address,
+    ) -> Result<String, Error> {
```
Comment (ed255) [inline bus-mapping/src/rpc.rs:127]: This endpoint returns the code as a hex string.  Could you decode the hex string and return a `Vec<u8>`?  This way we get the exact type we expect.

File: bus-mapping/src/rpc.rs [hunk 3]
```diff
@@ -199,4 +213,17 @@ mod rpc_tests {
...
         assert!(!trace_by_hash[0].struct_logs.is_empty());
     }
+
+    #[ignore]
+    #[tokio::test]
+    async fn test_get_contract_code() {
+        let transport = Http::new(Url::parse("http://localhost:8545").unwrap());
+        let prov = GethClient::new(transport);
+        let contract_address =
+            Address::from_slice(b"0xD5F110B3E81dE87F22Fa8c5e668a5Fc541c54E3d");
```
Comment (ed255) [inline bus-mapping/src/rpc.rs:223]: Can you confirm if you've this test locally successfully?
The method `Address::from_slice` is not supposed to take a hex string.  For that you can use `Address::from_str` or the `address` macro from `crate::address`.

**Review by ed255 — CHANGES_REQUESTED** [review]

**Comment by ed255** [comment]
> Hi @ed255 I drafted `eth_getCode` method issue #194. I also setup the test env with `ganache` which supports same RPC with geth. By indicating same mnemonic, we can produce same contract addresses and block hashes. You can check my code with `$ docker-compose up` and `cd bus-mapping && cargo test -- --ignored`.

I think ganache will be very limiting for our tests, because it doesn't support all the RPC endpoints that geth does.  Are you familiar with operating geth?  If so, do you think it's feasible to replace ganache with geth, so that the docker-compose starts a geth instance in development mode, and the js scripts deploy the contract there?  This way we could lay a path towards an integration test that will allow us to test complete flows.

File: bus-mapping/src/rpc.rs (hunk 3 above)
Comment (NoCtrlZ) [reply bus-mapping/src/rpc.rs]: Yeah I fixed.

File: bus-mapping/src/rpc.rs (hunk 2 above)
Comment (NoCtrlZ) [reply bus-mapping/src/rpc.rs]: I changed `String` to `Vec<u8>`.

**Comment by NoCtrlZ** [comment]
> If so, do you think it's feasible to replace ganache with geth, so that the docker-compose starts a geth instance in development mode, and the js scripts deploy the contract there?

Yeah it's piece of cake 🧁

> This way we could lay a path towards an integration test that will allow us to test complete flows.

I agree with that!

File: bus-mapping/src/rpc.rs (hunk 2 above)
Comment (ed255) [reply bus-mapping/src/rpc.rs]: I think you forgot to decode the hex!  You can do that very easily using this type https://docs.rs/ethers-core/0.5.6/ethers_core/types/struct.Bytes.html like:
```rust
let code: Bytes = ...
// ...
code.to_vec()
```

**Comment by NoCtrlZ** [comment]
There are two problems happening.
- `deployer` container can't access through host `geth`.
- web3 transaction is not confirmed.

```
➜  geth git:(issue/194) ✗ yarn deploy
yarn run v1.21.1
$ ts-node utils/deploy.ts
^C
➜  geth git:(issue/194) ✗ yarn deploy
yarn run v1.21.1
$ ts-node utils/deploy.ts
(node:39548) UnhandledPromiseRejectionWarning: Error: Returned error: already known
```
The assumption is that the transaction pool is not included to block.

**Comment by NoCtrlZ** [comment]
Okay!

**Pull request closed without merging**